### PR TITLE
ESQL: Track memory from values loaded from lucene

### PR DESCRIPTION
**Pull request by nik9000**
This adds memory tracking for values loaded from doc values and stored fields.

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/index/mapper/TestBlock.java [hunk 1]
```diff
@@ -192,8 +193,18 @@ public TestBlock appendOrd(int value) {
...
+    }
+
     private TestBlock add(Object value) {
         (currentPosition == null ? values : currentPosition).add(value);
         return this;
     }
+
+    @Override
+    public void close() {
+        // TODO assert that we close the test blocks
```
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/index/mapper/TestBlock.java:208]: I'll take a look at this one now.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlockBuilder.java [hunk 2]
```diff
@@ -140,8 +140,9 @@ protected final void ensureCapacity() {
             return;
         }
         int newSize = calculateNewArraySize(valuesLength);
-        adjustBreaker((long) (newSize - valuesLength) * elementSize());
+        adjustBreaker(newSize * elementSize());
         growValuesArray(newSize);
+        adjustBreaker(-valuesLength * elementSize());
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlockBuilder.java:145]: I made this change while trying to reenable one of the HeapAttack tests - specifically the too many mv fields one. It didn't get it passing, which is genuinely odd to me. Either way, I think it is more correct.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockUtils.java [hunk 3]
```diff
@@ -161,6 +161,7 @@ public static Block[] fromList(BlockFactory blockFactory, List<List<Object>> lis
     public static Block deepCopyOf(Block block, BlockFactory blockFactory) {
         try (Block.Builder builder = block.elementType().newBlockBuilder(block.getPositionCount(), blockFactory)) {
             builder.copyFrom(block, 0, block.getPositionCount());
+            builder.mvOrdering(block.mvOrdering());
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockUtils.java:164]: This was required to get some tests working. It meant I had to make the method a noop on unsupported block types but that seems ok.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperatorTests.java [hunk 4]
```diff
@@ -236,83 +229,73 @@ private void loadSimpleAndAssert(DriverContext driverContext, List<Page> input)
...
-                offset = mvDoubles.getFirstValueIndex(i);
-                for (int v = 0; v <= key % 3; v++) {
-                    assertThat(mvDoubles.getDouble(offset + v), equalTo(key / 123_456d + v));
-                }
-                if (key % 3 > 0) {
-                    assertThat(mvDoubles.mvOrdering(), equalTo(Block.MvOrdering.DEDUPLICATED_AND_SORTED_ASCENDING));
-                }
-            }
-        }
+        // for (Page p : results) {
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperatorTests.java]: Woop.s Let me get this back.

**Review by nik9000 — COMMENTED** [review]
I'm going to push an extra change here - I'll flip the tests from defaulting to an untracked BlockFactory to a tracked one. We want all tests to use a tracked one in soon enough, so let's get the default right.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlockBuilder.java (hunk 2 above)
Comment (dnhatn) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlockBuilder.java:145]: ++ this is more correct as we create a new array of `newSize` before copying.

**Review by dnhatn — APPROVED** [review]
LGTM.

**Pull request merged**